### PR TITLE
Improve Error Handling

### LINKED ISSUE #128: Improve handling of error response
When an error occurs, the server usually returns a specific error message, which is handled by `GenericPacket.read` method. However, methods that are calling #read are not always handling the error response (EmptyPacket for SMB1 and ErrorPacket for SMB2).

This issue is about reviewing every method that is processing response message from the remote host and make sure the error response is correctly handled.

Also, the `GenericPacket.read` method needs to rescue an `IOError` Exception class instead of `EOFError` to handle every possible issue. (`EOFError` is a sub-class of `IOError`).

### PR DESCRIPTION
**Pull request by chrisdlf**
This fixes #128

This PR updates the error handling logic by adding validation for packet structures.

## Changes
- Add valid? method to GenericPacket, ErrorPacket and EmptyPacket.
- Add packet validation for SMB1 and SMB2 packets.
- Update specs.
- Add some missing specs.
## Verification Steps
- [x] `bundle install`
- [x] Run all the examples in examples/ directory and verify errors are properly handled.
## `rake spec`
- [x] `rake spec`
- [x] VERIFY no failures

### CODE REVIEW

**Comment by chrisdlf** [comment]
The `initialize_instance` method has been removed from Requests files because the `command` field is now set in the parent class (`GenericPacket`). However, in Response files, the packet class still needs to set the SMB header `reply` flag when the initialization occurs. I could have also moved this to `GenericPacket`, but I thought the logic to detect if it is a response packet would be overkill.

Regarding the error messages, I wanted to keep some sort of unity and use the same type of message across all the validation errors (many existing error message were build this way). The message only includes the expected packet class name for simplicity.
I can change this and add both the `protocol` and `command` flags, since it is what the `valid?` method checks. Please, let me know and I will update this.

**Comment by jrobles-r7** [comment]
@chrisdlf could you add in the `protocol` and `command` information to the error messages, please?

**Comment by chrisdlf** [comment]
Sure! I added more details to InvalidPacket exceptions (commit 62275f1)

File: spec/lib/ruby_smb/client_spec.rb [hunk 1]
```diff
@@ -153,6 +154,94 @@
...
+
+    context 'with SMB2' do
+      let(:raw_response) { double('Raw response') }
+      let(:logoff_response) {
+        RubySMB::SMB2::Packet::LogoffResponse.new(smb_header: {:command => RubySMB::SMB2::Commands::LOGOFF} )
+      }
+      before :example do
+        allow(smb2_client).to receive(:send_recv).and_return(raw_response)
+        allow(RubySMB::SMB2::Packet::LogoffResponse).to receive(:read).and_return(logoff_response)
+        allow(smb1_client).to receive(:wipe_state!)
```
[the remark this answers is not in the document]
Comment (chrisdlf) [reply spec/lib/ruby_smb/client_spec.rb]: Bad copy/paste I'm afraid, thanks for pointing this out. Addressed in 9bf4ff9.

File: spec/lib/ruby_smb/smb1/tree_spec.rb [hunk 2]
```diff
@@ -414,6 +437,20 @@
       it 'returns the expected FindFileFullDirectoryInfo structures' do
         expect(tree.list).to eq([file_info1, file_info2])
       end
+
+      context 'when the response is not a valid Trans2 FindFirst2Response' do
```
[the remark this answers is not in the document]
Comment (chrisdlf) [reply spec/lib/ruby_smb/smb1/tree_spec.rb]: Addressed in 9bf4ff9.

**Comment by jrobles-r7** [comment]
Thanks for all of the improvements @chrisdlf !

**Pull request merged**